scalafmt-tests/src/test/resources/rewrite/RedundantParens.stat

```diff
@@ -31,3 +31,35 @@ object a {
       if (b ++ b).length >= 2
     } yield a
 }
+<<< compound statement
+object a {
+  def c(b: List[Int]): List[Int] =
+    for {
+      a <- b
+      if (a) || (a)
```

Comment:
Any idea why this does not have balanced parentheses? I'm a bit uneasy about this fix since I suspect the root cause of the problem is something different.

Reply:
This source code does have balanced parentheses.

However, the bug here is that RedundantParens only trims leading and trailing parentheses and in this case it would output:

```scala
  if a) || (a
```

I decided to go for a safe fix with this PR and check if the resulting code coming from RedundantParens (`innerSegment` above) has balanced parentheses. This may not handle all cases, but it seemed like an easy fix for the issue at hand.

One could probably be clever about actually removing the parentheses correctly, but I wanted to keep this change small to fix the bug and perhaps somebody else could improve RedundantParens in the future.

Reply:
Aaah, I see! Did you try using RewriteCtx.matchinParens? https://github.com/scalameta/scalafmt/blob/a9cc066127feca7384b03b516821e1840056d0db/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/Rewrite.scala#L14

You can use that map to guard against rewriting open/close pairs that don't match.

Reply:
Interesting - I did not know about that. I updated the code to use that instead.